Upgraded to pulumi v0.16.4 and node v11.2. Doing `pulumi new` failing on `npm install`

Here is a transcript.

[log.txt](https://github.com/pulumi/pulumi/files/2590469/log.txt)

There are a bunch of issues with gRPC on Node 11, which are all due to the fact that there are not prebuilt binaries for gRPC available.  The linked grpc issue points out that there are bugs in node-pre-gyp due to breaking changes in node 11 that can cause builds to race in some cases.

In this case, it doesn't look like a problem with the builds racing, but just the fact that the build of the gRPC native bits fail on this specific Windows machine.

For example, in the log, we the VC++ compiler crashed when trying to build boring SSL

```
..\deps\grpc\third_party\boringssl\crypto\asn1\a_enum.c(156): fatal error C1903: unable to recover from previous error(s); stopping compilation [C:\trypulumi\azure-stuff\node_modules\grpc\build\boringssl.vcxproj]
  Internal Compiler Error in C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\amd64\CL.exe.  You will be prompted to send an error report to Microsoft later.
  INTERNAL COMPILER ERROR in 'C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\amd64\CL.exe'
      Please choose the Technical Support command on the Visual C++
      Help menu, or open the Technical Support help file for more information
```

@dfinke Would it be possible for you to tell us what version of CL is being used here?  I'll try on my windows machine as well and see if I can get this building with the newest release visual c++ build tools.

Of course, it's not just Windows where we have problems.  I've personally hit issues where the c++ compiler on my linux machine issues some warnings on code which are then promoted to errors and I have to set `CXXFLAGS="-Wno-error"` before running `npm install`.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,9 @@
 
-## 0.16.5 (unreleased)
+## 0.16.6 (Unreleased)
+
+## 0.16.5 (Released Novemeber 16th, 2018)
+
+- Fix an issue where `pulumi plugin install` would fail on Windows with an access deined message.
 
 ## 0.16.4 (Released Novemeber 12th, 2018)
 

--- a/pkg/workspace/plugins.go
+++ b/pkg/workspace/plugins.go
@@ -152,44 +152,52 @@ func (info PluginInfo) Install(tarball io.ReadCloser) error {
 		return errors.Wrapf(err, "creating plugin directory %s", tempDir)
 	}
 
-	// Unzip and untar the file as we go.
-	defer contract.IgnoreClose(tarball)
-	gzr, err := gzip.NewReader(tarball)
-	if err != nil {
-		return errors.Wrapf(err, "unzipping")
-	}
-	r := tar.NewReader(gzr)
-	for {
-		header, err := r.Next()
-		if err == io.EOF {
-			break
-		} else if err != nil {
-			return errors.Wrapf(err, "untarring")
+	// Unzip and untar the file as we go. We do this inside a function so that the `defer`'s to close files happen
+	// before we later try to rename the directory. Otherwise, the open file handles cause issues on Windows.
+	err = (func() error {
+		defer contract.IgnoreClose(tarball)
+		gzr, err := gzip.NewReader(tarball)
+		if err != nil {
+			return errors.Wrapf(err, "unzipping")
 		}
+		r := tar.NewReader(gzr)
+		for {
+			header, err := r.Next()
+			if err == io.EOF {
+				break
+			} else if err != nil {
+				return errors.Wrapf(err, "untarring")
+			}
 
-		path := filepath.Join(tempDir, header.Name)
+			path := filepath.Join(tempDir, header.Name)
 
-		switch header.Typeflag {
-		case tar.TypeDir:
-			// Create any directories as needed.
-			if _, err := os.Stat(path); err != nil {
-				if err = os.MkdirAll(path, 0700); err != nil {
-					return errors.Wrapf(err, "untarring dir %s", path)
+			switch header.Typeflag {
+			case tar.TypeDir:
+				// Create any directories as needed.
+				if _, err := os.Stat(path); err != nil {
+					if err = os.MkdirAll(path, 0700); err != nil {
+						return errors.Wrapf(err, "untarring dir %s", path)
+					}
 				}
+			case tar.TypeReg:
+				// Expand files into the target directory.
+				dst, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR, os.FileMode(header.Mode))
+				if err != nil {
+					return errors.Wrapf(err, "opening file %s for untar", path)
+				}
+				defer contract.IgnoreClose(dst)
+				if _, err = io.Copy(dst, r); err != nil {
+					return errors.Wrapf(err, "untarring file %s", path)
+				}
+			default:
+				return errors.Errorf("unexpected plugin file type %s (%v)", header.Name, header.Typeflag)
 			}
-		case tar.TypeReg:
-			// Expand files into the target directory.
-			dst, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR, os.FileMode(header.Mode))
-			if err != nil {
-				return errors.Wrapf(err, "opening file %s for untar", path)
-			}
-			defer contract.IgnoreClose(dst)
-			if _, err = io.Copy(dst, r); err != nil {
-				return errors.Wrapf(err, "untarring file %s", path)
-			}
-		default:
-			return errors.Errorf("unexpected plugin file type %s (%v)", header.Name, header.Typeflag)
 		}
+
+		return nil
+	})()
+	if err != nil {
+		return err
 	}
 
 	return os.Rename(tempDir, finalDir)

--- a/tests/integration/integration_test.go
+++ b/tests/integration/integration_test.go
@@ -214,13 +214,12 @@ func TestStackOutputsJSON(t *testing.T) {
 	e.RunCommand("pulumi", "login", "--cloud-url", e.LocalURL())
 	e.RunCommand("pulumi", "stack", "init", "stack-outs")
 	e.RunCommand("pulumi", "up", "--non-interactive", "--skip-preview")
-	stdout, stderr := e.RunCommand("pulumi", "stack", "output", "--json")
+	stdout, _ := e.RunCommand("pulumi", "stack", "output", "--json")
 	assert.Equal(t, `{
   "foo": 42,
   "xyz": "ABC"
 }
 `, stdout)
-	assert.Equal(t, "", stderr)
 }
 
 // TestStackOutputsDisplayed ensures that outputs are printed at the end of an update

